dhis-2/dhis-services/dhis-service-node/src/main/java/org/hisp/dhis/node/serializers/CsvNodeSerializer.java

```diff
@@ -0,0 +1,193 @@
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.node.serializers;
+
+import java.io.OutputStream;
+import java.util.Date;
+import java.util.List;
+
+import org.hisp.dhis.node.AbstractNodeSerializer;
+import org.hisp.dhis.node.Node;
+import org.hisp.dhis.node.types.CollectionNode;
+import org.hisp.dhis.node.types.ComplexNode;
+import org.hisp.dhis.node.types.RootNode;
+import org.hisp.dhis.node.types.SimpleNode;
+import org.hisp.dhis.util.DateUtils;
+import org.springframework.context.annotation.Scope;
+import org.springframework.context.annotation.ScopedProxyMode;
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.dataformat.csv.CsvFactory;
+import com.fasterxml.jackson.dataformat.csv.CsvGenerator;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvSchema;
+import com.google.common.collect.Lists;
+
+/**
+ * @author Morten Olav Hansen <mortenoh@gmail.com>
+ */
+@Component
+@Scope( value = "prototype", proxyMode = ScopedProxyMode.INTERFACES )
+public class CsvNodeSerializer extends AbstractNodeSerializer
+{
+    private static final String[] CONTENT_TYPES = { "application/csv", "text/csv" };
+
+    private static final CsvMapper CSV_MAPPER = new CsvMapper();
+
+    private static final CsvFactory CSV_FACTORY = CSV_MAPPER.getFactory();
+
+    private CsvGenerator csvGenerator;
+
+    @Override
+    public List<String> contentTypes()
+    {
+        return Lists.newArrayList( CONTENT_TYPES );
+    }
+
+    @Override
+    protected void startSerialize( RootNode rootNode, OutputStream outputStream )
+        throws Exception
+    {
+        csvGenerator = CSV_FACTORY.createGenerator( outputStream );
+
+        CsvSchema.Builder schemaBuilder = CsvSchema.builder()
+            .setUseHeader( true );
+
+        // build schema
+        for ( Node child : rootNode.getChildren() )
+        {
+            if ( child.isCollection() )
+            {
+                if ( !child.getChildren().isEmpty() )
```

Comment:
could be combined - I think there is an intelliJ inspection that should flag this for you as well

Reply:
fixed